### ISSUE TITLE
Ever since update only returning one device?

### ISSUE THREAD

**Issue by marauder2k7** [kcat/openal-soft#751]
ever since the update to 1.22.2 this code

`if (ALFunction.alcIsExtensionPresent(NULL, "ALC_ENUMERATE_ALL_EXT")) {
          devices = (char *)ALFunction.alcGetString(NULL, ALC_ALL_DEVICES_SPECIFIER);
          defaultDeviceName = (char *)ALFunction.alcGetString(NULL, ALC_DEFAULT_ALL_DEVICES_SPECIFIER);
   }
   else
   {
       devices = (char *)ALFunction.alcGetString(NULL, ALC_DEVICE_SPECIFIER);
       defaultDeviceName = (char *)ALFunction.alcGetString(NULL, ALC_DEFAULT_DEVICE_SPECIFIER);
   }`
   
   is only returning one device into devices, before it was returning all the devices in the system.

**Comment by kcat** [comment]
How are you checking the returned names? For the `ALC_ALL_DEVICES_SPECIFIER` and `ALC_DEVICE_SPECIFIER` strings, the names are separated by null characters, and is finished by a blank name.

**Comment by marauder2k7** [comment]
> 

i can share a link to the source code: https://github.com/TorqueGameEngines/Torque3D/blob/development/Engine/source/sfx/openal/aldlist.cpp

I am working on a refactor of the entire sound system atm and want to go 100% openal-soft and use its latest features. The idea of this code is to split the OpenAL Soft on Speakers (' deviceName ') so that in the options menu it just shows the device name for people

**Comment by kcat** [comment]
You're modifying the returned device list:
```c++
char* openFind = dStrchr(devices, '(');
if (openFind)
{
   devices = openFind + 1;
   char* closeFind = dStrchr(devices, ')');
   if (closeFind)
      (*closeFind) = '\0';
```
The returned string is `const char*`, so you must not modify it. In this case, it's causing the last character of the name (which is `)`) to be overwritten with a null char, so the subsequent `dStrlen` counts up to what would have been `)`, with the next character being the real null char for the name instead of the start of the next name, which causes the loop to break thinking it found an empty name.

Be aware also that the format of the device string isn't guaranteed. It's based on the name returned by the system, but different systems may format device names differently. Even Windows isn't guaranteed to name devices the same way in the future. There's also no guarantee the device name won't itself contain another pair of `(`/`)` chars, which would break the logic. It's best to not assume anything about the device names returned by OpenAL Soft, other than they're user-readable UTF-8 strings.

**Comment by marauder2k7** [comment]
yeah i just noticed my mistake, rather stupid moment after coding for hours lol sorry about that. 

Just curious is there a faster way to load all of openal-soft? i noticed the router class but unsure if its meant to be used internally or can it be used by our game engine?

**Comment by kcat** [comment]
What's being slow to load?

The router is a separate DLL that manages multiple OpenAL implementations, so you can link directly to one OpenAL and use the standard functions, and get access to all implementations on the system, be it OpenAL Soft, Rapture3D, Creative's wrapper driver, a hardware driver, etc. They'll all be shown as separate devices for enumeration, and which device you open will then determine which implementation is in use and what extensions will be available. Since you're going 100% OpenAL Soft for its newest features, it won't benefit you.
